### PR TITLE
Replace kaldiio with omniio for Kaldi ark/scp I/O

### LINKED ISSUE #6529: About kaldiio license compatibility with ESPnet production use
The existing discussion in #5549 raised this question, but it was closed as stale without a clarification. I am opening a new issue because the concern remains relevant to current ESPnet versions.

The current [kaldiio license](https://github.com/nttcslab-sp/kaldiio/blob/master/LICENSE) appears to restrict use to internal evaluation and limit redistribution, while [ESPnet still lists kaldiio as a dependency](https://github.com/espnet/espnet/blob/master/pyproject.toml).

I am concerned that this creates uncertainty for publicly distributed software.

### PR DESCRIPTION
**Pull request by sw005320**
## What did you change?

Follow-up to #6547. Replaces `kaldiio` with [`omniio`](https://pypi.org/project/omniio/), which exposes the same API and writes byte-identical archives, and removes the optional-dependency machinery #6547 built — omniio is MIT-licensed, so none of it is needed.

**Net effect: 44 files, +122/−221.** The diff is mostly deletions.

- Every call site is an ordinary `from omniio import kaldi as kaldi_io`. `espnet2/utils/kaldiio_utils.py` and its `import_kaldiio()` helper are gone, along with the deferred calls at all 17 sites.
- `omniio>=0.1.0` is in `[project].dependencies`. No extra, no installer, no Makefile target, no CI install line.
- `ci/no_redistribute.txt` loses its only entry, as the file itself anticipated: *"Slated for replacement; drop this line when that lands."* The mechanism stays, empty — it is generic, and kaldiio coming off the list is the point.
- The `try/except ImportError` guards around `omniio.interface` and `omniio.text.read` in the speechlm loaders are gone too. They predate this branch and had nothing left to guard; `audio_read` was called unchecked at its use site anyway, so the guard only turned a clear `ImportError` into a `TypeError`.
- The prebuilt image no longer smoke-tests `import omniio.kaldi`. That check earned its place while omniio arrived out of band; an ordinary dependency needs no more proof of import than `numpy` does. k2 keeps its, for the reason its comment gives.

## Why did you make this change?

#6547 closed #6529 by making `kaldiio` opt-in, but the underlying problem is unchanged for anyone who actually needs Kaldi features: its licence is an NTT evaluation-only agreement whose section 4(b) forbids distributing, transferring or reproducing the software. That is why it was the sole entry in `ci/no_redistribute.txt` and the reason `ci/report_unlicensed.py` exists.

omniio is MIT-licensed, so the dependency can simply be declared and every code path can import it directly.

## Cost, stated plainly

`pip install espnet` now also installs **av, Pillow and zstandard — about 23 MB**. `pyarrow` and `tqdm` are already pulled in by `datasets`, so they are not new. That is the price of dropping the indirection; flagging it because a new required dependency affects every user.

## Compatibility

Archives are byte-identical, asserted in both directions in omniio's own CI against `kaldiio` as a reference implementation: `FM`/`DM` and `FV`/`DV` matrices and vectors, `CM`/`CM2`/`CM3` compressed matrices for all seven `compression_method` values, `std::vector<int32>` alignments, bare-RIFF `WaveHolder` entries, the `AUDIO`-framed extended layout used by `format_wav_scp.py`, text arks, scp offsets and `segments`.

Two documented divergences, both in degenerate cases: `kSpeechFeature` compression of matrices with fewer than five rows follows Kaldi rather than `kaldiio`'s wrapping arithmetic, and the fixed-range methods clip out-of-range input instead of letting it wrap. Neither affects a real feature matrix.

omniio ships `omniio/tools/kaldi/PROVENANCE.md`, which records that its `kaldi` module was written from Kaldi's own Apache-2.0 sources and from reading bytes back, that `kaldiio`'s source was not read, and the measurements behind that — zero shared internal helper names, zero shared comments, 3.7 % similarity over the tree.

I ran ESPnet's own Kaldi code paths against the published package: `file_writer_helper`/`file_reader_helper` compressed and uncompressed, `ReadHelper(segments=...)`, `CMVN` via both the `mat` and `ark` paths, an extended ark written the way `format_wav_scp.py` writes it and read back the way `kaldi_loader` reads it, and the TTS x-vector path — Kaldi-native `FV` archives as `nnet3-xvector-compute` writes them, the two shapes `extract_spk_embed.py` produces, and text-mode vectors.

## Checks

`flake8` (over `test utils doc espnet2`), `isort` and `black` clean on every touched file; `ci/check_ci_image_config.py` passes all eleven invariants.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by sw005320** [comment]
@coderabbitai Thanks — one applied and widened, one overtaken by events.

**Lazy import in `compute_vad.py` and `mfa_format.py`.** Right, and the same shape held in two more files, so all four are fixed in e63f064:

| | Kaldi I/O is reached only when |
|---|---|
| `compute_vad.py` | `wav.scp` holds a pipe entry |
| `convert_token2json.py` | likewise |
| `kaldi_to_whisper.py` | likewise |
| `mfa_format.py` | a `segments` file exists, or a source is not `.wav` |

The common path in each — a plain `wav.scp` of `.wav` files — never touches it, yet the module-level import made omniio mandatory to start the script. They now call `import_kaldi_io()` at the use site, which is what `format_wav_scp.py` already did for the identical pipe branch.

The scripts still importing it directly are the ones that genuinely cannot do anything without it: `dump_codec`, `dump_audio_tokens`, `extract_spk_embed` and its parallel variant, `concatenate_feature` and `prepare_s2t_format` all read or write ark/scp as their entire purpose. That is the split #6547 set out, and these four were simply on the wrong side of it.

Verified with an interpreter that has no omniio installed: no module-level omniio import remains in any of the four, `import_kaldi_io()` appears only inside the conditional branches, and the error a user hits is the actionable one:

```
`omniio` is not installed. It is an optional dependency of ESPnet, used only for
Kaldi ark/scp I/O. Please install it with `pip install "espnet[omniio]"` or
`pip install omniio`.
```

**Pinning the omniio revision.** This one no longer applies — the review ran against the previous revision of the PR, where omniio was installed from a Git URL in four places because it was not published and `ci/check_ci_image_config.py`'s `check_no_direct_references` correctly refuses a `name @ url` dependency in `pyproject.toml`.

omniio 0.1.0 is now on PyPI, so 271bee0 removed all four of those paths along with `tools/installers/install_omniio.sh` and the `omniio.done` Makefile target. The dependency is an ordinary extra:

```toml
omniio = ["omniio>=0.1.0"]
```

No mutable Git ref is fetched anywhere, and the version constraint follows the same convention as every other dependency in this file. `check_no_direct_references` is now satisfied by construction rather than worked around.

File: egs2/TEMPLATE/asr1/pyscripts/audio/compute_vad.py [hunk 1]
```diff
@@ -7,11 +7,11 @@
 from pathlib import Path
 from typing import Optional, Tuple
 
-import kaldiio
 import numpy as np
 import resampy
 import soundfile
 import soundfile as sf
+from omniio import kaldi as kaldi_io
```
[the remark this answers is not in the document]
Comment (sw005320) [reply egs2/TEMPLATE/asr1/pyscripts/audio/compute_vad.py:14]: Handled, then deliberately reverted — flagging that here so the thread is not misleading.

`e63f064` did exactly what this asked: `compute_vad.py` and `mfa_format.py` moved to `import_kaldi_io()` at the use site, and I found the same shape in two more files (`convert_token2json.py` and `kaldi_to_whisper.py`) and fixed those too.

`063038b9` then undid it. The premise no longer holds:

> Without `omniio`, their top-level imports prevent all other paths from starting.

omniio is now in `[project].dependencies`, not an extra, so there is no "without omniio" state to protect. It is MIT-licensed, which is the whole reason the optional-dependency machinery from #6547 could go — `espnet2/utils/kaldi_io_utils.py` and `import_kaldi_io()` no longer exist, and all 17 call sites are plain module-level imports.

So the current `from omniio import kaldi as kaldi_io` at `compute_vad.py:14` and `mfa_format.py:35` is intentional, and the same lazy-import argument would now apply equally to `numpy` or `soundfile`.

The cost of that choice, stated in the PR description: `pip install espnet` also pulls av, Pillow and zstandard, about 23 MB. pyarrow and tqdm are already required via `datasets`.

**Comment by sw005320** [comment]
@wanchichen, please review this PR

**Comment by sw005320** [comment]
The `lm` integration failure is fixed upstream; this branch now requires **omniio 0.1.1**.

## What broke

```
File "egs2/mini_an4/lm1/local/prepare_lm_data.py", line 22, in unit2cjk
    return chr(int("4e00", 16) + int(unit))
ValueError: invalid literal for int() with base 10: '5.0'
```

omniio's text-ark writer formatted every value with `repr(float(v))`, so the k-means pseudo-labels that `perform_kmeans.sh` writes with `ark,t:` came out with decimal points:

```
kaldiio:  utt1  [ 5 12 3 0 ]
omniio:   utt1  [ 5.0 12.0 3.0 0.0 ]
```

and `prepare_lm_data.py` parses those tokens back with `int()`. Alignments have the same shape.

## Two more found while fixing it

- **Negative values in binary integer vectors were read unsigned.** Kaldi stores `-5` as `fb ff ff ff`; omniio wrote those bytes correctly and byte-identically, but read them back as `4294967291`. Any Kaldi alignment containing a negative value.
- **Text arks with CRLF line endings were misparsed.** A matrix is told from a vector by whether a newline follows the `[`, and `\r` defeated that, so a 2×2 matrix came back as a 4-element vector with the wrong dtype and no error.

Both released in 0.1.1 (wavlab-speech/omniio#14). Verified against a clean `pip install omniio==0.1.1`:

```
text ark:        'utt1  [ 5 12 3 0 ]'
int() で読める:  [5, 12, 3, 0]
負値の往復:      [-5  0  7]
CRLF 行列:       (2, 2) float32
```

**No ESPnet code changes** — the writer was what was wrong, and the recipe was right. The only change here is the version floor.

Worth saying: this is exactly the kind of defect the unit tests could not catch. omniio's text-ark tests only ever used float arrays; running the thing inside a real recipe is what found it. omniio now parametrises those tests over `int32`, `int64`, `uint8`, `float32` and `float64` on both the write and read sides, and asserts byte-identity against `kaldiio` for each.

File: espnet2/utils/kaldi_io_utils.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ESPnet previously used ``kaldiio`` here.  It was replaced because its license
+restricts redistribution, which is a problem for anyone publicly distributing
+software that depends on ESPnet (see https://github.com/espnet/espnet/issues/6529);
+``omniio`` is MIT-licensed and writes byte-identical archives.
+"""
+
+import importlib
+from types import ModuleType
+
+OMNIIO_URL = "git+https://github.com/wavlab-speech/omniio.git"
```
[the remark this answers is not in the document]
Comment (sw005320) [reply espnet2/utils/kaldi_io_utils.py]: Valid at `a0de553d6`, resolved at head by a different route — and worth saying why the suggested remedy would not have been the right one.

At head `fd4e5392f` there is no git URL left in any of the four paths. `pyproject.toml` has

```toml
# Kaldi ark/scp I/O. Replaced kaldiio, whose license restricts redistribution.
"omniio>=0.1.1",
```

and `ci/install.sh` and `ci/install_macos.sh` are no longer in this PR's diff at all. OmniIO was published to PyPI as `omniio` 0.1.1 (uploaded 2026-09-04 19:06 via trusted publishing), so the installation path is now an ordinary version requirement rather than any kind of Git fetch, pinned or not.

**Pinning the URL to a commit SHA would not have worked as a fix**, which is the part worth recording. PyPI refuses any distribution whose metadata contains a PEP 508 direct reference, pinned or not:

```
400 Can't have direct dependency: name @ git+https://...
```

That is what kept espnet off PyPI from `v.202604` to `v.202609` — five months of failed uploads, with `twine check` reporting PASSED on the rejected artifact the whole time. `ci/check_ci_image_config.py` gained a `check_no_direct_references` invariant in #6629 precisely so a `git+` in `pyproject.toml` fails on the pull request instead of at the release tag, so the pinned version of this finding would now be rejected by CI as well.

The general shape, for future reviews of this repository: when a dependency can only be had from a fork, the fix is to publish that fork to PyPI under its own name — as was done for `omniio` here, and for `espnet-ctc-segmentation`, `espnet-g2p-en` and `espnet-s3prl` in #6611 — not to pin the reference.

**Comment by wanchichen** [comment]
The PR looks good to me, hard to spot any notable issues. If there are any bugs, its probably upstream in omniio on another missed edge case, like the one above.

**Comment by sw005320** [comment]
@wanchichen, can you run the test with some of your Kaldi arks in your working directories?
You can scan and sample them to test.
I found several issues in my working directories, and we need more tests.
(see https://github.com/wavlab-speech/omniio/pull/16)

**Comment by sw005320** [comment]
@wanchichen — packaged the sweep so you don't have to reconstruct it. Two commands:

```bash
pip install "omniio>=0.1.2" kaldiio
curl -sLO https://raw.githubusercontent.com/wavlab-speech/omniio/main/scripts/compare_with_kaldiio.py
python compare_with_kaldiio.py /path/to/your/dump
```

It reads every archive under those directories with **both** omniio and kaldiio and compares values and dtypes. Read-only, and exits non-zero if anything mismatched.

**Look at the by-format table, not the total.** "1420 files, 0 mismatches" is worth very little if all 1420 were the same kind of record:

```
=== by format (this is the part that matters) ===
  WaveHolder (RIFF)                     1200   mismatched 0
  CM (compressed, per-column)            220   mismatched 0
```

If a format you care about isn't in that list, it wasn't exercised — point it somewhere that has some. Defaults sample 25 entries per `.scp`; `--max-per-file 0` reads everything.

**If your data has speed perturbation, add `--allow-pipes`.** That's where both bugs were. `sox` piping to stdout can't seek back to patch its WAV header, so `sph2pipe … | sox … speed 0.9 |` leaves a placeholder size — omniio before 0.1.2 refused those outright, which killed `file_reader_helper(filetype="sound")` and the `kaldi_ark` loader on sp data. The flag is off by default because reading such an entry means running the command in it; with it on, only allow-listed programs writing to stdout are run (never `gunzip archive.gz` or `sox a.wav b.wav` — those delete or overwrite), stages are exec'd directly with no shell, and output is capped under a timeout.

One caveat that cost me a false alarm: don't compare by running the pipe twice. A `sox` stage without `-R` dithers non-repeatably, so the samples differ no matter what the libraries do. The script runs it once and hands the same bytes to both.

`--out mismatches.tsv` gives one line per problem. Please paste the by-format table either way — a clean run on a corpus I don't have is a real result.

**Comment by sw005320** [comment]
@wanchichen — gentle nudge on this one. No rush if you're busy, but it's the last thing holding the PR: everything else is green and the branch is ~37 commits behind master now, so I'd rather not let it drift much further.

If you have a dump directory handy, it's about two minutes of your time:

```bash
pip install "omniio>=0.1.2" kaldiio
curl -sLO https://raw.githubusercontent.com/wavlab-speech/omniio/main/scripts/compare_with_kaldiio.py
python compare_with_kaldiio.py /path/to/your/dump
```

Just paste the by-format table, clean or not. The reason I'm asking rather than relying on CI: both bugs found so far were in data that no fixture produces — archives written years ago by Kaldi, `sox` and `sph2pipe`. Your working directories are a different sample from mine, which is the whole value.

And if you don't have anything suitable to point it at, just say so and I'll merge on what we have.

**Comment by wanchichen** [comment]
Sorry for the delay, here are my results. All positive. I tested on some OWSM datasets (pre-dumped to .ark), raw wavs, sph2pipe (CHIME), and with `sox` + `ffmpeg`.

```
===== [1] owsm train_3.2 wav.scp (2000-line reservoir sample, 23 datasets) =====

  AUDIO (extended)                       500   mismatched 0


===== [2] chime4 tr05_simu_noisy (.wav) + iwslt test_clean (ark) + chime4 tr05_orig_clean (sph2pipe) =====
  WaveHolder (RIFF)                      300   mismatched 0
  AUDIO (extended)                       300   mismatched 0

sph2pipe pipe entries (chime4 tr05_orig_clean)
sampled=300  compared=300  mismatched=0  pipe_errors=0

===== [3] pipes: speed-perturb + sox + ffmpeg =====
  pipe (sox)                             210   mismatched 0
  pipe (ffmpeg)                           90   mismatched 0

chained sph2pipe | sox speed 0.9 (espnet pipe-source form)
compared=30 mismatched=0
example sp0_9-pipe-0: kaldiio=(rate=16000,(116346,) int16) omniio=(rate=16000,(116346,) int16) equal=True

############################################################
# TOTAL: 1430 entries checked, 0 MISMATCHES.
# Formats: AUDIO(ext) ark, WaveHolder(RIFF) .wav, sph2pipe SPHERE,
#          sox pipes (incl. speed-perturb), ffmpeg pipes.
############################################################
```
Many of my dumps were deleted last year, so can't test on too many datasets.
